### PR TITLE
Revamp extension loading

### PR DESCRIPTION
**Pull request by redref**
Use python3 capabilities to load extensions.

Maintain compatibility with previous extensions.

### CODE REVIEW

File: pkr/__main__.py [hunk 1]
```diff
@@ -20,18 +20,21 @@
 def main():
     """Main function"""
     try:
-        cli_args = get_parser().parse_args()
+        parser = get_parser()
+        cli_args = parser.parse_args()
 
         # Setting the log mode
-        debug = cli_args.__dict__.pop('debug')
+        debug = getattr(cli_args, 'debug')
```
Comment (Maillol) [inline pkr/__main__.py:27]: I don't understand the `__dict__.pop` or the getattr. The usually and the KISS method is
`cli_args.debug`

File: pkr/__main__.py [hunk 2]
```diff
@@ -20,18 +20,21 @@
...
         log.set_debug(debug)
 
-        func = cli_args.__dict__.pop('func')
+        func = getattr(cli_args, 'func')
         func(cli_args)
     except KardInitializationException as exc:
         log.write(str(exc))
         return 1
     except Exception as exc:  # pylint: disable=W0703
+        if '--debug' in sys.argv or '-d' in sys.argv:
```
Comment (Maillol) [inline pkr/__main__.py:36]: If you need to do that, that means you except that an error can be occurs will the argument parsing. 
is only possible if you use a poorly coded  custom type in the `argparse.ArgumentParser.add_argument` method.

For me this code is useless.

File: pkr/ext/__init__.py [hunk 3]
```diff
@@ -54,41 +57,52 @@ class Extensions(object):
...
 
     @staticmethod
-    def _wrap_call(extension, method_name, *args, **kwargs):
-        method = getattr(extension.plugin(), method_name, None)
+    def _wrap_call(name, extension, method_name, *args, **kwargs):
+        method = getattr(extension, method_name, None)
+        base_method = getattr(ExtMixin, method_name, None)
 
-        if method is None:
+        if method is None or method == base_method:
```
Comment (Maillol) [inline pkr/ext/__init__.py:100]: here you can use `is` insteadof `==` 
We don't know if method and base_method have the same value but if method is base _method.

File: pkr/__main__.py (hunk 1 above)
Comment (Kimamisa) [reply pkr/__main__.py:27]: I think the __dict__.pop was to remove the attribute at the same time

File: pkr/__main__.py (hunk 2 above)
Comment (redref) [reply pkr/__main__.py:36]: I agree, but we have such cases.

File: pkr/ext/__init__.py (hunk 3 above)
Comment (redref) [reply pkr/ext/__init__.py:100]: Updated.

File: pkr/__main__.py [hunk 4]
```diff
@@ -20,18 +20,21 @@
...
+        parser = get_parser()
+        cli_args = parser.parse_args()
 
         # Setting the log mode
-        debug = cli_args.__dict__.pop('debug')
+        debug = getattr(cli_args, 'debug')
         log.set_debug(debug)
 
-        func = cli_args.__dict__.pop('func')
+        func = getattr(cli_args, 'func')
```
[the remark this answers is not in the document]
Comment (redref) [reply pkr/__main__.py:30]: Tried to simplify. But we have cases where the parser return exception.

File: pkr/cli/parser.py [hunk 5]
```diff
@@ -103,7 +103,13 @@ def _create_kard(args):
...
         extra_features = args.features
         if extra_features is not None:
-            extras.update({'features': extra_features.split(',')})
+            extra_features = extra_features.split(',')
+            if 'features' in extras:
+                extra_features.extend(
+                    x for x in extras['features'] if x not in extra_features)
+                extras['features'] = extra_features
+            else:
+                extras.update({'features': extra_features})
```
Comment (redref) [inline pkr/cli/parser.py:112]: Add the ability to mix features in command line (-f) and features in meta file (-m)

File: pkr/__main__.py (hunk 2 above)
Comment (Kimamisa) [reply pkr/__main__.py:25]: ok. Maybe add the explanation as a comment in that case.

File: pkr/cli/parser.py (hunk 5 above)
Comment (Kimamisa) [reply pkr/cli/parser.py:111]: :+1:

Comment (Kimamisa) [reply pkr/cli/parser.py:111]: maybe add it through a set, to avoid having the same feature twice

Comment (redref) [reply pkr/cli/parser.py:111]: set is messing with the order, I prefer the deterministic way, but it is covered here.

File: pkr/driver/__init__.py [hunk 6]
```diff
@@ -3,9 +3,22 @@
...
+    'compose': 'docker_compose',
+    'kubernetes': 'k8s',
+}
 
 def load_driver(driver_name):
     """Return the loaded driver"""
-    return DriverManager(namespace='drivers', name=driver_name).driver
+    if driver_name in DRIVER_MAPPING:
+        driver_name = DRIVER_MAPPING[driver_name]
+    module = import_module('pkr.driver.{}'.format(driver_name))
```
Comment (Kimamisa) [inline pkr/driver/__init__.py:19]: Maybe a try / catch would be interesting here

File: pkr/cli/parser.py (hunk 5 above)
Comment (redref) [reply pkr/cli/parser.py:111]: deterministic to be able to compare meta files actually.

File: pkr/driver/__init__.py (hunk 6 above)
Comment (Kimamisa) [reply pkr/driver/__init__.py:19]: Also, you should consider using the optional package argument of this function.
Smth like : import_module(driver_name, 'pkr.driver')

Comment (Kimamisa) [reply pkr/driver/__init__.py:19]: cf https://docs.python.org/3/library/importlib.html#importlib.import_module

File: pkr/ext/__init__.py [hunk 7]
```diff
@@ -54,41 +57,52 @@ class Extensions(object):
     extensions
     """
 
-    def __init__(self, features):
+    def __init__(self, features=None):
         super(Extensions, self).__init__()
-        self.features = set(features)
+        self.features = features
```
Comment (Kimamisa) [inline pkr/ext/__init__.py:62]: do you make sure somewhere that you do not have any duplicate ?

File: pkr/__main__.py (hunk 2 above)
Comment (redref) [reply pkr/__main__.py:28]: Added message. Closing for now. Feel free to reopen if needed.

File: pkr/cli/parser.py (hunk 5 above)
Comment (redref) [reply pkr/cli/parser.py:120]: With new version, we should be able to handle order / unicity / merging. Will resolve the other conversation on the subject, most of the code is here.

Comment (redref) [reply pkr/cli/parser.py:120]: pkr kard update will need to be addressed, but I propose to handle it in a subsequent PR targeted to pkr kard update itself.

File: pkr/kard.py [hunk 8]
```diff
@@ -197,29 +197,30 @@ def set_meta(kard, extra):
         merge(kard.driver.get_meta(extra, kard), kard.meta)
 
         # Prevent duplication of features in case of update
-        kard.meta['features'] = list(set(kard.meta['features']))
+        kard.meta['features'] = list(kard.meta['features'])
```
Comment (Kimamisa) [inline pkr/kard.py:200]: in that case, you can remove the whole line...

File: pkr/kard.py [hunk 9]
```diff
@@ -197,29 +197,30 @@ def set_meta(kard, extra):
...
 
         # Extensions
         kard.extensions.setup(extra, kard)
 
         # We add all remaining extra to the meta
         merge(extra, kard.meta)
-        kard.meta['features'] = list(set(kard.meta['features']))
+        kard.meta['features'] = list(kard.meta['features'])
```
Comment (Kimamisa) [inline pkr/kard.py:207]: here too..

File: pkr/kard.py (hunk 8 above)
Comment (redref) [reply pkr/kard.py]: Done !

File: pkr/kard.py (hunk 9 above)
Comment (redref) [reply pkr/kard.py]: Done !

File: pkr/driver/__init__.py (hunk 6 above)
Comment (redref) [reply pkr/driver/__init__.py]: Fixed by calling pkr.driver.* on pkr.driver package

**Pull request merged**